### PR TITLE
Build OS X

### LINKED ISSUE #1: Build iris on OS X
Right now we had to skip due to https://github.com/conda-forge/staged-recipes/pull/280#issuecomment-206377121

### PR DESCRIPTION
**Pull request by ocefpaf**
This is not working. Just sending the PR so we can have a log of the issue.

Closes #1

### CODE REVIEW

**Comment by jakirkham** [comment]
Awesome. Thanks for doing this @ocefpaf.

**Comment by jakirkham** [comment]
While it would be hard to pick my favorite features from conda-forge. If this is not number 1, it is at least in the top 3. Being able to trivially share build logs and reference issues like this is simply fantastic.

**Comment by ocefpaf** [comment]
Problem one: the solver chooses to downgrade `netcdf4` to 1.2.3 and, by not using `1.2.4` that is pinned against `libnetcdf 4.3*`, it upgrades `libnetcdf`  to `4.4` and breaks everything.

Possible solutions: Delete the older build and hope for the best. Over pin by added `libnetcdf-4.3*` to iris.

```shell
===== testing package: iris-1.9.2-np110py27_0 =====
import: 'iris'
Traceback (most recent call last):
  File "/Users/travis/miniconda3/conda-bld/test-tmp_dir/run_test.py", line 25, in <module>
    import iris
  File "/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/iris/__init__.py", line 112, in <module>
    import iris.fileformats
  File "/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/iris/fileformats/__init__.py", line 35, in <module>
    from . import netcdf
  File "/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/iris/fileformats/netcdf.py", line 38, in <module>
    import netCDF4
  File "/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/netCDF4/__init__.py", line 3, in <module>
    from ._netCDF4 import *
ImportError: dlopen(/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/netCDF4/_netCDF4.so, 2): Library not loaded: @rpath/./libnetcdf.7.dylib
  Referenced from: /Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/netCDF4/_netCDF4.so
  Reason: image not found
TESTS FAILED: iris-1.9.2-np110py27_0
```

**Comment by ocefpaf** [comment]
Problem two: if we overcome problem one `iris` still does not work on `Python 3`. See https://github.com/conda-forge/staged-recipes/pull/280#issuecomment-206377121 and https://github.com/SciTools/conda-recipes-scitools/pull/140

**Comment by jakirkham** [comment]
Yeah, we should delete the old `netcdf`s. They don't pin `hdf5` either, which is a huge problem. See this [line]( https://travis-ci.org/conda-forge/iris-feedstock/jobs/128075367#L2141 ). This can also be seen by inspecting the packages in `conda-forge`.

**Comment by ocefpaf** [comment]
That is the way conda is now and we don't really have an option :unamused: 

I would love to see another behavior though. My Linux distro add rules to:

- never downgrade a build number unless it is a force install;
- and to always maximize the versions of the **specified** requirements.

The latter means that `netcdf4` there would never be downgraded and the former is to ensure that we would never get a broken package as build number bumps are, 99% of the time, fixing packaging issues.
When those rules cannot be satisfied the it would at least ask:

```shell
The specified dependencies are forcing packages `x,y,z` to downgrade. Are you sure [y/N]
```

However, I don't really want to start another conda solver thread. I just want to say that there is a better behavior and we don't need to re-invent the wheel.

**Comment by jakirkham** [comment]
Is it possible that we are pinned to `netcdf` 1.2.3 in some place? If that is the case, maybe we can release a one-off of `netcdf` 1.2.3 with a higher build number (than already present) and add the right pinnings in it.

**Comment by ocefpaf** [comment]
> Is it possible that we are pinned to `netcdf` 1.2.3 in some place?

No.

> If that is the case, maybe we can release a one-off of `netcdf` 1.2.3 with a higher build number (than already present) and add the right pinnings in it.

No need. I removed the previous unpinned builds to avoid giving conda the bad choice of downgrading `libnetcdf`. That should get us building `Python 2.7`, but then we are still stuck with problem 2 from above.

PS: I hate to delete old builds as that is a kick in the stomach for those, like me, worried about provenance and reproducibility. However, until we fix our tools we (which I know @pelson would prefer) we don't really have a choice but to workaround this issue.

**Comment by jakirkham** [comment]
> No need. I removed the previous unpinned builds...

Ok.

> That should get us building `Python 2.7`, but then we are still stuck with problem 2 from above.

Just so I am clear, this is not a problem on Linux or Windows?

In any event, this would be better than the current situation. I have a few things that won't build on Windows or some Python versions on Windows due to weird issues.

> I hate to delete old builds...

Fair. We can try upping the build number first with proper fixes on the given release in the future.

> However, until we fix our tools we (which I know @pelson would prefer)...

Let's wait and see how their meeting goes on Monday. Though there are some new people being added to the mix. So it might take a couple of meetings to get something concrete.

> ...we don't really have a choice but to workaround this issue.

Trying the above strategy first would at least be one choice we can make. There is no guarantee that it solves it though.

**Comment by ocefpaf** [comment]
> Just so I am clear, this is not a problem on Linux or Windows?

By this you mean problem 1 or problem 2? Either way both are only on OS X, but probably due to different reasons.

@pelson I am puzzled that [PY34 NP111](https://travis-ci.org/conda-forge/iris-feedstock/jobs/128098325) passed! Maybe that help us debug this...

**Comment by pelson** [comment]
Just catching up, so there may be a number of comments coming. First though:


> and to always maximize the versions of the specified requirements.

conda-build is now doing this in master I believe https://github.com/conda/conda-build/pull/907

**Comment by jakirkham** [comment]
Yep, was following that. Also, @pelson and @ocefpaf, @msarahan was planning on doing a beta release of `conda-build` today (assuming all goes well). So, we should be able to get this functionality if we start using the beta. This would be a good time to raise any pressing bugs that you would like fixed. This should help us clear `curl`, which has been a major bottleneck. It will also fix some issues with shebangs on Mac (I believe).

**Comment by pelson** [comment]
> No need. I removed the previous unpinned builds to avoid giving conda the bad choice of downgrading libnetcdf. That should get us building Python 2.7, but then we are still stuck with problem 2 from above.

Looking good, thanks @ocefpaf:

```
$ conda execute conda-forge.github.io/scripts/list_deps.py conda-forge --package netcdf --platform osx-64

Fetching package metadata: ..
netcdf-cxx4 4.2.1 0 osx-64         : libnetcdf
netcdf-cxx4 4.2.1 1 osx-64         : libnetcdf 4.3*
netcdf-fortran 4.4.3 2 osx-64      : krb5, libgcc, libnetcdf 4.3*
netcdf4 1.2.4 np110py27_1 osx-64   : hdf5 1.8.15*, libnetcdf 4.3*, numpy 1.10*, python 2.7*, setuptools
netcdf4 1.2.4 np110py34_1 osx-64   : hdf5 1.8.15*, libnetcdf 4.3*, numpy 1.10*, python 3.4*, setuptools
netcdf4 1.2.4 np110py35_1 osx-64   : hdf5 1.8.15*, libnetcdf 4.3*, numpy 1.10*, python 3.5*, setuptools
netcdf4 1.2.4 np111py27_1 osx-64   : hdf5 1.8.15*, libnetcdf 4.3*, numpy 1.11*, python 2.7*, setuptools
netcdf4 1.2.4 np111py34_1 osx-64   : hdf5 1.8.15*, libnetcdf 4.3*, numpy 1.11*, python 3.4*, setuptools
netcdf4 1.2.4 np111py35_1 osx-64   : hdf5 1.8.15*, libnetcdf 4.3*, numpy 1.11*, python 3.5*, setuptools
```

> PS: I hate to delete old builds as that is a kick in the stomach for those, like me, worried about provenance and reproducibility. However, until we fix our tools we (which I know @pelson would prefer) we don't really have a choice but to workaround this issue.

Agreed. Removing builds breaks reproducibility. On the other hand, maintaining the existing builds breaks usability. I've been considering the idea of having 2 conda-forge channels within the anaconda.org conda-forge user - a "main" and a "legacy". We could avoid deleting anything from "legacy", but remove anything which isn't immaculate in "main". There are some details there, but at least (with a bit of effort) users can *completely* reproduce an environment from a few months ago.

**Comment by pelson** [comment]
Looks like an iris issue now, rather than a conda/conda-build issue:

```
copying lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb -> build/lib.macosx-10.5-x86_64-3.4/iris/fileformats/_pyke_rules
/Users/travis/miniconda3/envs/_build/bin/python tools/generate_std_names.py etc/cf-standard-name-table.xml build/lib.macosx-10.5-x86_64-3.4/iris/std_names.py
writing [iris.fileformats._pyke_rules.compiled_krb]/fc_rules_cf_fc.py
writing [iris.fileformats._pyke_rules.compiled_krb]/compiled_pyke_files.py
Traceback (most recent call last):
  File "setup.py", line 229, in <module>
    'clean_source': CleanSource},
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/site-packages/setuptools-20.7.0-py3.4.egg/setuptools/command/install.py", line 61, in run
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/distutils/command/install.py", line 539, in run
    self.run_command('build')
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/distutils/command/build.py", line 126, in run
    self.run_command(cmd_name)
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "setup.py", line 180, in run
    MakePykeRules._pyke_rule_compile()
  File "setup.py", line 134, in _pyke_rule_compile
    knowledge_engine.engine(iris.fileformats._pyke_rules)
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/site-packages/pyke/knowledge_engine.py", line 110, in __init__
    target_package.load(self, **kws)
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/site-packages/pyke/target_pkg.py", line 366, in load
    self.do_by_ext('load', target_filename, engine, load_flags)
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/site-packages/pyke/target_pkg.py", line 265, in do_by_ext
    return getattr(self, "%s_%s" % (prefix, ext))(filename, *args)
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/site-packages/pyke/target_pkg.py", line 374, in load_py
    self.load_module(module_path, target_filename, engine)
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/site-packages/pyke/target_pkg.py", line 407, in load_module
    module = import_(module_path)
  File "/Users/travis/miniconda3/envs/_build/lib/python3.4/site-packages/pyke/target_pkg.py", line 463, in import_
    mod = __import__(modulename)
ImportError: No module named 'iris.fileformats._pyke_rules.compiled_krb'
Command failed: /bin/bash -x -e /Users/travis/build/conda-forge/iris-feedstock/recipe/build.sh
```

**Comment by ocefpaf** [comment]
Yep. All I could do was to get you to Problem 2 :wink: 

https://github.com/conda-forge/iris-feedstock/pull/2#issuecomment-217189398

**Comment by jakirkham** [comment]
> Looks like an iris issue now, rather than a conda/conda-build issue...

I would agree, but...

> I am puzzled that [PY34 NP111](https://travis-ci.org/conda-forge/iris-feedstock/jobs/128098325) passed!

Feels funny doesn't it? :/

**Comment by pelson** [comment]
Lol. Thank for reminding me of my poor memory @ocefpaf 👍 
I'll see if I can reproduce locally.

**Comment by ocefpaf** [comment]
> Lol. Thank for reminding me of my poor memory @ocefpaf 👍 

BTW that is the last piece on the ioos stack holding us to Python 2.7. So no pressure :wink:

**Comment by jakirkham** [comment]
> BTW that is the last piece on the ioos stack holding us to Python 2.7.

You're lucky you only have one piece.

**Comment by ocefpaf** [comment]
> You're lucky you only have one piece.

Lucky? No! Many, many PRs to many places :smile:

**Comment by pelson** [comment]
"Luck Is What Happens When Preparation Meets Opportunity" - Seneca

😉

**Comment by pelson** [comment]
P.S. I'm able to reproduce the problem locally.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -42,6 +45,7 @@ test:
     imports:
         - iris
         - iris.fileformats.pp_packing  # [not win]
+        - iris.fileformats._pyke_rules.compiled_krb  # [not win]
```
Comment (ocefpaf) [inline recipe/meta.yaml]: @pelson On your PR you forgot to re-render. So I absorbed you changes here.
Also, this import was failing on Windows there. Not sure if that is supposed to work on Windows.

Comment (pelson) [inline recipe/meta.yaml]: Yes it must. Thanks for the reminder. The same magic line will be needed in the bld.bat.

File: recipe/build.sh [hunk 2]
```diff
@@ -15,4 +15,10 @@ echo "udunits2_path = $PREFIX/lib/libudunits2.${EXT}" >> $SITECFG
...
 rm -rf lib/iris/tests/results lib/iris/tests/*.npz
 
-$PYTHON -sE setup.py --with-unpack build_ext "-I${PREFIX}/include" "-L${PREFIX}/lib" "-R${PREFIX}/lib" install --single-version-externally-managed --record record.txt
+$PYTHON -sE setup.py --with-unpack std_names \
+    build_ext "-I${PREFIX}/include" "-L${PREFIX}/lib" "-R${PREFIX}/lib" \
+    install --single-version-externally-managed --record record.txt
+
+# Without this line, the OS X build fails reproducibly with
+# https://github.com/conda-forge/iris-feedstock/pull/2#issuecomment-217468453.
+$PYTHON -c "import iris.fileformats.netcdf; iris.fileformats.netcdf._pyke_kb_engine()"
```
Comment (pelson) [inline recipe/build.sh]: This needs to come into the bld.bat.

File: recipe/meta.yaml (hunk 1 above)
Comment (ocefpaf) [inline recipe/meta.yaml]: Or we could apply the patch to `OS X` only, no?

File: recipe/build.sh (hunk 2 above)
Comment (ocefpaf) [inline recipe/build.sh]: What if we just apply the patch to OS X instead?

File: recipe/meta.yaml (hunk 1 above)
Comment (pelson) [inline recipe/meta.yaml]: Yes.

Comment (pelson) [inline recipe/meta.yaml]: We don't even need to make the changes to build.sh conditional

Comment (ocefpaf) [inline recipe/meta.yaml]: OK.

**Comment by pelson** [comment]
🎉  - bravo @ocefpaf. 😄

**Comment by jakirkham** [comment]
> Lucky? No! Many, many PRs sent to many places :smile:

> "Luck Is What Happens When Preparation Meets Opportunity" - Seneca

The issue is not all of this software is stuff I maintain. Some of it (sadly) is not even on GitHub so it is pretty rough. We use a lot of breaking edge stuff so Python 3 is not a priority for the developers/alogrithm writers, publishing is. Even if I do help somebody convert something over to Python 3, it seems to be at my expense. So, I think this is more a matter of walking in my shoes ( 👞 ) for a month to understand what I am saying here.

**Comment by ocefpaf** [comment]
Thanks for for investigating and creating the patch @pelson! I will be running my very first 100% Python 3.5 production env tomorrow! :tada:

**Comment by lexual** [comment]
Interestingly, I got caught by this or something similar, building py36 packages from miniconda2 install.
Switching to miniconda3 seemed to resolve my issue ;)

**Pull request merged**